File: Final_Project/Code/Project_4/Bai_toan_5/Exercises/4/CSS_bt5_ex4.py
class Graph:
    def __init__(self, n):
        # n: so dinh
        # adj: danh sach ke
        self.n = n
        self.adj = [[] for _ in range(n)]

    def add_edge(self, u, v):
        # Them canh (u,v)
        self.adj[u].append(v)
        self.adj[v].append(u)

    def generate_complete(self):
        # Sinh do thi day du K_n
        for i in range(self.n):
            for j in range(i + 1, self.n):
                self.add_edge(i, j)

    def generate_bipartite(self, p, q):
        # Sinh do thi luong phan day du K_{p,q}
        for i in range(p):
            for j in range(p, p + q):
                self.add_edge(i, j)

# Vi du su dung
if __name__ == "__main__":
    g = Graph(4)
    g.generate_complete()
    print("Complete graph edges:", 
          [(u, v) for u in range(g.n) 
                   for v in g.adj[u] if u < v])
    g = Graph(5)
    g.generate_bipartite(2, 3)
    print("Bipartite graph edges:", 
          [(u, v) for u in range(g.n) 
                   for v in g.adj[u] if u < v])

File: Final_Project/Code/Project_4/Bai_toan_5/Exercises/4/CSS_bt5_ex4.cpp
#include <iostream>
#include <vector>
using namespace std;

class Graph {
public:
    int n;                       // so dinh
    vector<vector<int>> adj;     // danh sach ke

    Graph(int vertices) : n(vertices) {
        adj.resize(n);
    }

    void add_edge(int u, int v) {
        // Them canh (u,v)
        adj[u].push_back(v);
        adj[v].push_back(u);
    }

    void generate_complete() {
        // Sinh do thi day du K_n
        for (int i = 0; i < n; ++i) {
            for (int j = i + 1; j < n; ++j) {
                add_edge(i, j);
            }
        }
    }

    void generate_bipartite(int p, int q) {
        // Sinh do thi luong phan day du K_{p,q}
        for (int i = 0; i < p; ++i) {
            for (int j = p; j < p + q; ++j) {
                add_edge(i, j);
            }
        }
    }
};

int main() {
    Graph g(4);
    g.generate_complete();
    cout << "Complete graph edges: ";
    for (int u = 0; u < g.n; ++u) {
        for (int v : g.adj[u]) {
            if (u < v) cout << "(" << u << "," << v << ") ";
        }
    }
    cout << endl;
    return 0;
}